### DISCUSSION TITLE
Requests not processed as expected

### DISCUSSION THREAD

**Discussion by SaiUpadhyayula** [vladimir-bukhtoyarov/bucket4j#500] [Q&A]
Hello,

**Context**:
We are trying to implement Rate Limiting for our custom Webhook platform at our company, we have a system publishing requests and these requests will be sent to the consumers who subscribe to the publisher.

Our consumers now want to have some of these messages ratelimited as they can't handle the huge spike of requests from publisher at once, so we are implement server side rate limiting on our side using Bucket4J.

We allow the consumer to configure a value of 10-200 rps. Here is how is our setup:

- All the incoming messages from publisher are sent to a RabbitMQ queue
- We have a scheduled listener created using Quartz manually polling this queue for every 1 second, reads N number of messages from the queue and tries to send these messages to the consumer. The call to the consumer is protected by the bucket4J logic, that is configured with the below configuration:

```
            int requestsPerSecond = subscription.getRequestsPerSecond();
            Bandwidth limit = BandwidthBuilder.builder()
                    .capacity(requestsPerSecond)
                    .refillGreedy(requestsPerSecond, Duration.ofSeconds(1))
                    .build();
            Supplier<BucketConfiguration> supplier = () -> BucketConfiguration.builder()
                    .addLimit(limit)
                    .build();
            var proxy = proxyManager.builder()
                    .withOptimization(Optimizations.batching())
                    .build(key, supplier);
```

We are using Redis with Lettuce Client to store the state of the bucket, this is configured as follows:

```
        RedisClient redisClient = redisClient();
        StatefulRedisConnection<String, byte[]> redisConnection = redisClient
                .connect(RedisCodec.of(StringCodec.UTF8, ByteArrayCodec.INSTANCE));

        return LettuceBasedProxyManager.builderFor(redisConnection)
                .withClientSideConfig(ClientSideConfig.getDefault())
                .build();
```

And the call to the consumer is surrounded by the below logic:

```
        BucketProxy bucketProxy = bucketResolver.resolveBucket(event, subscription);
        ConsumptionProbe consumptionProbe = bucketProxy.tryConsumeAndReturnRemaining(1);
        if (consumptionProbe.isConsumed()) {
            log.info("Tokens consumed, remaining messages {}", consumptionProbe.getRemainingTokens());
            decoratorWrapper.test(() -> sendNotification(event, attemptTracker.isInitial()));
        } else {
            var exchangeQueueName = queueName + SenderRegistry.EXCHANGE_EXT;
            log.info("Message with eventId {} is throttled, sending it back to the original queue {}", event.getEventId(), exchangeQueueName);
            rabbitMqGenericProducer.persistEventByExchange(exchangeQueueName, event);
        }
```

As the consumer can update the number of requests at any time, we are changing the bucket configuration as below:

```
                 Bandwidth limit = BandwidthBuilder.builder()
                        .capacity(subscription.getRequestsPerSecond())
                        .refillGreedy(subscription.getRequestsPerSecond(), Duration.ofSeconds(1))
                        .build();
                BucketConfiguration bucketConfiguration = BucketConfiguration.builder()
                        .addLimit(limit)
                        .build();
                bucketRequestTracker.bucketProxy()
                        .replaceConfiguration(bucketConfiguration, TokensInheritanceStrategy.PROPORTIONALLY);
```

**Problem**:

We observed that when we try to set the requestsPerSecond to any value greater than 60, we always see that the max requests sent per second do not exceed 45-50.

I am not sure what is the bottleneck and why we are not able to reach the required number of requests.

Any value < 60, we are exactly reaching that number of requests.

We are testing this configuration by simulating the load using the below command:

`ab -n 1000 -p <payload> -T "application/json" -c 10 -A "$username:$password" "$URL"`

Appreciate any help to fix the issue, Thanks

**Comment by vladimir-bukhtoyarov** [comment]
Hello @SaiUpadhyayula 

Current integration with Redis is built on top of Compare&Swap paradigm. It is not a problem when requests are well distributed across multiple keys, but problems happen in case of significant contention on same(like in your case). Looks that 45-50rps is maximum for your env.

You have two options:
* Try to use Hazelcast/Ignite/Infinispan/Coherence instead of Redis, its integrations with Bucket4j are built on top of high-perfomant EntryProcessor paradigm.
* If switching to another technology is not accaptable then try optimizations that described there https://github.com/bucket4j/bucket4j/issues/479#issuecomment-2070919340

**Reply by SaiUpadhyayula** [reply]
Thanks for your prompt response, do you think this will work if we try to integrate mariaDB as the backend ? Unfortunately we don't have out of the box support for any of the above mentioned technologies in our organization.

**Reply by vladimir-bukhtoyarov** [reply]
I do not think that MariaDB will show significantly better results.

I see that you are already use batching optimizations:
```java
var proxy = proxyManager.builder()
                    .withOptimization(Optimizations.batching())
                    .build(key, supplier);
```

I have missed that while was posting first answer. You should to change your code a bit - `bucketProxy.resolveBucket` should cache reference to bucketProxy, because [as it mentioneed in javadocs](https://github.com/bucket4j/bucket4j/blob/8.12/bucket4j-core/src/main/java/io/github/bucket4j/distributed/proxy/RemoteAsyncBucketBuilder.java#L63) batching optimization is applied if and only if requests are applied to the same bucketProxy object.

**Reply by SaiUpadhyayula** [reply]
@vladimir-bukhtoyarov I didn't provide the full code for the resolveBucket() method previously, I am actually storing the bucketProxy in a map to retrieve it based on the key and use it to consume the token.

```
public BucketProxy resolveBucket(Event event, Subscription subscription) {
        var key = String.format("%s_%s_%s", event.getEventType(), event.getSubscriptionApplicationName(),
                subscription.getSubscriptionId());
        BucketRequestTracker bucketRequestTracker = bucketConfigurationMap.get(subscription.getSubscriptionId());
        if (bucketRequestTracker == null) {
            int requestsPerSecond = subscription.getRequestsPerSecond();
            Bandwidth limit = BandwidthBuilder.builder()
                    .capacity(requestsPerSecond)
                    .refillGreedy(requestsPerSecond, Duration.ofSeconds(1))
                    .build();
            Supplier<BucketConfiguration> supplier = () -> BucketConfiguration.builder()
                    .addLimit(limit)
                    .build();
            var proxy = proxyManager.builder()
                    .withOptimization(Optimizations.batching())
                    .build(key, supplier);
            BucketRequestTracker newBucketRequestTracker = new BucketRequestTracker(proxy, requestsPerSecond);
            bucketConfigurationMap.put(subscription.getSubscriptionId(), newBucketRequestTracker);
            return proxy;
        } else {
            // Check if number of requests is changed, if yes then replace the configuration of bucket
            if (!bucketRequestTracker.requestsPerSecond().equals(subscription.getRequestsPerSecond())) {
                Bandwidth limit = BandwidthBuilder.builder()
                        .capacity(subscription.getRequestsPerSecond())
                        // We refill the bucket every second, as we are providing throttling
                        // on per second basis
                        .refillGreedy(subscription.getRequestsPerSecond(), Duration.ofSeconds(1))
                        .build();
                BucketConfiguration bucketConfiguration = BucketConfiguration.builder()
                        .addLimit(limit)
                        .build();
                bucketRequestTracker.bucketProxy()
                        .replaceConfiguration(bucketConfiguration, TokensInheritanceStrategy.PROPORTIONALLY);
            }
            return bucketRequestTracker.bucketProxy();
        }
    }
```

**Reply by vladimir-bukhtoyarov** [reply]
Code looks good. How many concurrent servers were used in the test? I am asking about of amount of servers behind "$URL" parameters of "ab" command. If it was only one server that problem should be in something else, not in CAS, because there is no contention.

Also, I forgot to ask how did you detect that that the max requests sent per second do not exceed 45-50. It is very interesting to know the rate of rejected requests when `consumptionProbe.isConsumed()==false`

Also it is interesting to measure the latency of interaction with bucket, in order to be sure that there is no problem in perfomance.

**Reply by SaiUpadhyayula** [reply]
@vladimir-bukhtoyarov We are using only one server as of now, and we are using a service virtualization mock endpoint to simulate the call to the consumer, we can see the calls made from our application in a Kibana dashboard as it's basically a mock call to the server, that's why I was able to say exactly the number.

One thing I noticed is the requests are not rejected at all for any number > 50 rps, I can see for eg: for 30rps, I can see that the tokens are emptied and the requests are rejected.

I am logging the remaining number of tokens using the consumptionProbe.getRemainingTokens() after each consumption, I see that they are just decreased by 2-5 requests before they are refilled again.

I also tried to run the http call to the consumer in a different thread but I still didn't find any difference in the throughput. This call is is taking around 70-80 ms.

Average time take for the interaction with the bucket is min.10 and max.26 ms, we can take an average of 12-13 ms.

In our dev environment I am performing this test using an xs instance of redis, which has just 128 MB of memory and storage, I upgraded this to a 2G memory and storage instance, but I also didn't find a difference.

Thanks a lot for your support.

**Reply by vladimir-bukhtoyarov** [reply]
@SaiUpadhyayula 

I have found one bug in batching optimization when it behaves wrong if count of parallel requests to same bucket > 4. Try to set system property **bucket4j.batching.merging-threshold=1000000** and repeat the test

**Reply by SaiUpadhyayula** [reply]
@vladimir-bukhtoyarov Unfortunately this setting is also not working as expected, I created a small POC that replicates the setup I have in my project.

https://github.com/SaiUpadhyayula/bucket4j-throttling-issue

Can you please check if it's something I am missing

When I run this example on my cloud environment I see the same behaviour as my actual application.

**Reply by vladimir-bukhtoyarov** [reply]
I have no time to investigate third-party code like spring-boot or rabbitmq, that independs from `bucket4js-redis`. I have wrote [isolated benchmark for bucket4j-redis](https://github.com/bucket4j/bucket4j/blob/master/bucket4j-examples/src/test/java/example/distributed/optimizers/batching/RedisBatchingExample.java#L77), it looks that all goes fine - any meaningfull rate can be achived. Use it to create self-contained example of problem. If you unable to reproduce problem in well-isolated code, then it explicitly means that problem in something else, like default settings of spring-boot-rabbitmq, rabbitmq-client, spring-boot shcedulers or something else outside `bucket4j-redis`.

**Reply by SaiUpadhyayula** [reply]
@vladimir-bukhtoyarov I understand, thanks again for your response
